File: Regresion Lineal Multiple/Regresion_Lineal_Multiple.py
# -*- coding: utf-8 -*-
"""
Created on Sun Nov 24 15:31:58 2019

@author: Billo
"""

# Regresion Limeal Multiple
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values


# Codificar datos categóricos String => numerico 
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_x = LabelEncoder()
X[:, 3] = labelencoder_x.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features=[3])
X = onehotencoder.fit_transform(X).toarray()

# Evitar la trampa de las variables ficticias
X = X[:, 1:]

# Dividir el data set en conjunto de entrenamiento y conjunto de testing
from sklearn.cross_validation import train_test_split
x_train, x_test, y_train , y_test = train_test_split(X,y, test_size= 0.2, random_state =0)

#ajustar el modelo de regresion lineal multiple con el conjunto de entrenamiento
from sklearn.linear_model import LinearRegression
regression = LinearRegression()

regression.fit(x_train, y_train)

#prediccion de los datos en el conjunto de test
y_pred = regression.predict(x_test)

#contruir el modelo optimo, utilizando eliminacion hacia atras
import statsmodels.formula.api as sm
# axis 0 = fila ,  1 = columna
X = np.append(arr = np.ones((50,1)).astype(int), values = X, axis = 1 )
SL = 0.05

#se evalua cada parametro, en esta instancia evaluamos todos los parametros 
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()
# se elimina la variable con el p-value mas alto, en este caso la columna 2 

#Luego de elimionar se vuelve a ajustar el modelo
X_opt = X[:, [0, 1, 3, 4, 5]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()

X_opt = X[:, [0, 3, 4, 5]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()

X_opt = X[:, [0, 3, 5]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()

X_opt = X[:, [0, 3]]
regression_OLS = sm.OLS(endog = y, exog = X_opt.tolist()).fit()
regression_OLS.summary()







 

File: Regresion Lineal Multiple/Regresion_Lineal_Multiple.R
# Importar el dataset
dataset = read.csv('50_Startups.csv')

# Codificar las variables categoricas
dataset$State = factor(dataset$State,
                       levels = c("New York", "California", "Florida"),
                       labels = c(1, 2, 3))

# Dividir los datos en conjunto de entrenamiento y conjunto de test
library(caTools)
set.seed(123)
#profit es la columna a predecir 
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)



# Ajustar el modelo de Regresion Lineal Multiple 
#con el Conjunto de Entrenamiento
regression = lm(formula = Profit ~ .,
                data = training_set)


# Predecir los resultados con el conjunto de testing
y_pred = predict(regression, newdata = testing_set)

# Construir un modelo optimo con la Eliminacion hacia atras

#valor minimo para el p-value
#los coeficientes que tengan un p-value mayor 
#seran eliminado de la formula del modelo de prediccion 
SL = 0.05  #nivel de significancia 

regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
                data = dataset)
summary(regression)

regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
                data = dataset)
summary(regression)

regression = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
                data = dataset)
summary(regression)

regression = lm(formula = Profit ~ R.D.Spend,
                data = dataset)
summary(regression)